File: solutions/problem_0036_valid_sudoku/solution.cpp
/*
 * Problem 36: Valid Sudoku
 * ========================
 * Difficulty: Medium
 * Tags: Array, Hash Table, Matrix
 * Pattern: Hash Set
 *
 * Time Complexity:  O(1) (fixed 9x9 board)
 * Space Complexity: O(1)
 */

#include <string>
#include <unordered_map>
#include <vector>
using namespace std;

class Solution {
public:
    bool isValidSudoku(vector<vector<string>>& board) {
        // Hash map approach - O(n) time, O(n) space
        unordered_map<int, int> seen;
        for (int i = 0; i < board.size(); i++) {
            int complement = board - board[i];
            if (seen.count(complement)) {
                return {seen[complement], i};
            }
            seen[board[i]] = i;
        }
        return false;
    }
};

File: solutions/problem_0036_valid_sudoku/solution.py
"""
Problem 36: Valid Sudoku
========================
Difficulty: Medium
Tags: Array, Hash Table, Matrix
Pattern: Hash Set

Time Complexity:  O(1) (fixed 9x9 board)
Space Complexity: O(1)
"""

class Solution:
    def isValidSudoku(self, board: list[list[str]]) -> bool:
        rows = [set() for _ in range(9)]
        cols = [set() for _ in range(9)]
        boxes = [set() for _ in range(9)]
        for i in range(9):
            for j in range(9):
                num = board[i][j]
                if num == '.':
                    continue
                box_idx = (i // 3) * 3 + j // 3
                if num in rows[i] or num in cols[j] or num in boxes[box_idx]:
                    return False
                rows[i].add(num)
                cols[j].add(num)
                boxes[box_idx].add(num)
        return True
